client/hwlib/src/collectors/cpuinfo.rs

```diff
@@ -0,0 +1,197 @@
+
+            let parts: Option<(_, _)> = line.split(':').collect_tuple();
+            if let Some((key, value)) = parts {
+                let key = key.trim();
+                let value = value.trim();
+                if key == "processor" {
+                    cores_count += 1;
+                }
+                attributes.insert(key, value);
+            }
+        }
+
+        let arch = std::env::consts::ARCH;
+        let speed_str = attributes.get("cpu MHz").copied();
+        let speed = CpuSpeed::try_from(speed_str)?.m_hz();
+
+        let platform = arch.to_string();
+        let model = attributes
+            .get("Model")
+            .or_else(|| attributes.get("model name"))
+            .unwrap_or(&arch)
+            .to_string();
+
+        let cpu_type = attributes.get("vendor_id").unwrap_or(&arch).to_string();
+        let model_number = attributes
+            .remove("cpu family")
+            .unwrap_or_default()
+            .to_string();
+        let model_version = attributes.remove("model").unwrap_or_default().to_string();
+        let model_revision = attributes
+            .remove("stepping")
+            .unwrap_or_default()
+            .to_string();
+        let cache = match attributes.remove("cache size") {
+            Some(data) => parse_cache_size(data)?,
+            None => -1,
+        };
+
+        let bogomips = match attributes.remove("bogomips") {
+            Some(data) => parse_bogomips(data)?,
+            None => -1,
+        };
+
+        Ok(CpuInfo {
+            platform,
+            cores_count,
+            cpu_type,
+            model,
+            model_number,
+            model_version,
+            model_revision,
+            cache,
+            bogomips,
+            speed,
+        })
+    }
+}
+
+pub struct CpuFrequency {
+    pub m_hz: u64,
+}
+
+impl CpuFrequency {
+    /// CPU frequency in MHz
+    pub const MHZ: CpuFrequency = CpuFrequency::from_m_hz(1);
+    pub const KHZ: CpuFrequency = CpuFrequency::from_k_hz(1000);
+
+    /// Read max CPU frequency from file and parse it in MHz as it's done in checkbox.
+    /// https://github.com/canonical/checkbox/blob/3789fdd/providers/resource/bin/cpuinfo_resource.py#L56-L63
+    pub fn from_file(max_cpu_frequency_filepath: &'static str) -> Result<Self> {
+        let raw_freq = fs::read_to_string(max_cpu_frequency_filepath)?;
+        let k_hz: u64 = raw_freq.trim().parse()?;
+        Ok(Self::from_k_hz(k_hz))
+    }
+
+    /// Create a CpuFrequency from a frequency in kHz
+    pub const fn from_k_hz(freq_khz: u64) -> Self {
+        Self {
+            m_hz: freq_khz / 1000,
+        } // Convert kHz to MHz
+    }
+
+    /// Create a CpuFrequency from a frequency in MHz
+    pub const fn from_m_hz(freq_mhz: u64) -> Self {
+        Self { m_hz: freq_mhz }
+    }
+}
+
+struct CpuSpeed(f64);
+
+impl CpuSpeed {
+    /// Get the frequency in MHz as a rounded integer
+    fn m_hz(&self) -> u64 {
+        self.0.round() as u64
+    }
```

Comment:
for creating a request, we use `CpuFrequency` object, and we need it to be an integer (and it is)
`CpuSpeed` is used to construct a `CpuInfo` object the same way checkbox does it. The speed field from that struct is not used currently, but I added it for future use, like when we don't have the data for constructing `CpuFrequency` object. Since `Processor` frequency is an integer, we want it to return an integer as well

Reply:
I see, thanks for the clarification!